In [ ]:
%load_ext autoreload
%autoreload 2

import sys
sys.path.insert(0, '')

# B6 (partie 2) — Auto-encodeur & détection d'anomalies

Reprend les jeux prétraités en [partie 1](b6_deep_learning_donnees.md) (`screw/*.bin.gz`) pour concevoir,
entraîner et évaluer un auto-encodeur convolutionnel de détection d'anomalies.

In [2]:
from pathlib import Path

import numpy as np

from src.vision.data import load_bin_gz

DATA_DIR = Path('deep-learning/screw')
IMG_SIZE = 128

X_train_norm = load_bin_gz(DATA_DIR / 'train_norm.bin.gz', IMG_SIZE)
test_good_norm = load_bin_gz(DATA_DIR / 'test_good_norm.bin.gz', IMG_SIZE)
test_defects_norm_stacked = load_bin_gz(DATA_DIR / 'test_defects_norm.bin.gz', IMG_SIZE)

print('Train (normé)  :', X_train_norm.shape, X_train_norm.dtype)
print('Test saines    :', test_good_norm.shape)
print('Test défauts   :', test_defects_norm_stacked.shape)

Train (normé)  : (256, 128, 128, 1) float32
Test saines    : (41, 128, 128, 1)
Test défauts   : (119, 128, 128, 1)


Conception de l'auto-encodeur

In [3]:
from src.vision.autoencoder import build_autoencoder

INPUT_SHAPE = X_train_norm.shape[1:]  # (128, 128, 1)

autoencoder, latent = build_autoencoder(INPUT_SHAPE, name='autoencodeur_screw')
autoencoder.summary()

Model: "autoencodeur_screw"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ image (InputLayer)              │ (None, 128, 128, 1)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d (Conv2D)                 │ (None, 64, 64, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 32, 32, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 16, 16, 16)     │         9,232 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_transpose                │ (None, 32, 32, 64)     │        16,448 │
│ (Conv2DTranspose)               │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_transpose_1              │ (None, 64, 64, 32)     │        32,800 │
│ (Conv2DTranspose)               │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_transpose_2              │ (None, 128, 128, 1)    │           513 │
│ (Conv2DTranspose)               │                        │               │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 77,809 (303.94 KB)

 Trainable params: 77,809 (303.94 KB)

 Non-trainable params: 0 (0.00 B)

In [4]:
input_size = np.prod(INPUT_SHAPE)
latent_size = np.prod(latent.shape[1:])
ratio = latent_size / input_size

print(f'Taille entrée       : {input_size} valeurs {INPUT_SHAPE}')
print(f'Taille goulot       : {latent_size} valeurs {tuple(latent.shape[1:])}')
print(f'Ratio goulot/entrée : {ratio:.3f}  (compression ×{1 / ratio:.1f})')

Taille entrée       : 16384 valeurs (128, 128, 1)
Taille goulot       : 4096 valeurs (16, 16, 16)
Ratio goulot/entrée : 0.250  (compression ×4.0)


**Commentaire.** Le modèle compte **77 809 paramètres** (tous entraînables), essentiellement dans les
6 couches de convolution/déconvolution — pas de couche dense, donc pas d'explosion de paramètres malgré
la taille des images (128×128). Le décodeur utilise un noyau de taille **4** (multiple du stride 2, au lieu
de 3) pour éviter l'artefact de damier ("checkerboard") typique des `Conv2DTranspose`, ce qui explique le
léger surplus de paramètres par rapport à un noyau 3×3.

Le **goulot d'étranglement** est la sortie de la 3ᵉ convolution : `16×16×16 = 4096` valeurs, contre
`128×128×1 = 16384` valeurs en entrée → **ratio de 0.25** (compression ×4). Ce ratio est volontairement
resserré via le nombre de filtres (32 → 64 → **16**, et non croissant jusqu'au bout) : avec un goulot proche
de la taille d'entrée, le modèle pourrait apprendre une quasi-identité et reconstruire aussi bien les
défauts que les pièces saines, ce qui ferait disparaître le signal d'anomalie recherché.

*Mise à jour* : le goulot était initialement à 8 canaux (ratio 0.125, ×8). Élargi à 16 canaux (ratio 0.25,
×4) pour réduire le flou de reconstruction observé même sur des saines (bruit de fond qui noyait le signal
des petits défauts). Gain mesuré : AUROC 0.978 → 0.982, faux négatifs 20 → 16 (à seuil centile-95 constant)
— amélioration réelle mais modeste, cohérente avec l'hypothèse plutôt qu'un changement radical.

# Étape 2

Entrainement du model, on entraine (enregistrement via MLFlow) un model

In [5]:
from src.vision.autoencoder import train_autoencoder

EPOCHS = 80 # 1 EPOCH => toutes les images sont passées et la backprop (rétropropagation) est passée
BATCH_SIZE = 16
PATIENCE = 4
MIN_DELTA = 1e-3

history = train_autoencoder(
    autoencoder, X_train_norm, latent.shape[1:], ratio,
    experiment_name='b6_deep_learning_autoencodeur_16_16_16',
    run_name='autoencoder_conv_screw_ssim',
    epochs=EPOCHS, batch_size=BATCH_SIZE, patience=PATIENCE, min_delta=MIN_DELTA,
)

Epoch 1/80


13/13 - 5s - 373ms/step - loss: 0.2868 - val_loss: 0.2575


Epoch 2/80


13/13 - 5s - 377ms/step - loss: 0.2372 - val_loss: 0.2140


Epoch 3/80


13/13 - 5s - 368ms/step - loss: 0.1957 - val_loss: 0.1789


Epoch 4/80


13/13 - 4s - 330ms/step - loss: 0.1674 - val_loss: 0.1501


Epoch 5/80


13/13 - 5s - 408ms/step - loss: 0.1431 - val_loss: 0.1375


Epoch 6/80


13/13 - 5s - 356ms/step - loss: 0.1362 - val_loss: 0.1329


Epoch 7/80


13/13 - 5s - 348ms/step - loss: 0.1314 - val_loss: 0.1282


Epoch 8/80


13/13 - 5s - 390ms/step - loss: 0.1267 - val_loss: 0.1231


Epoch 9/80


13/13 - 4s - 345ms/step - loss: 0.1210 - val_loss: 0.1158


Epoch 10/80


13/13 - 4s - 338ms/step - loss: 0.1129 - val_loss: 0.1064


Epoch 11/80


13/13 - 5s - 380ms/step - loss: 0.1034 - val_loss: 0.0976


Epoch 12/80


13/13 - 6s - 490ms/step - loss: 0.0946 - val_loss: 0.0885


Epoch 13/80


13/13 - 9s - 715ms/step - loss: 0.0860 - val_loss: 0.0831


Epoch 14/80


13/13 - 11s - 843ms/step - loss: 0.0799 - val_loss: 0.0742


Epoch 15/80


13/13 - 18s - 1s/step - loss: 0.0731 - val_loss: 0.0692


Epoch 16/80


13/13 - 11s - 871ms/step - loss: 0.0690 - val_loss: 0.0662


Epoch 17/80


13/13 - 13s - 1s/step - loss: 0.0653 - val_loss: 0.0629


Epoch 18/80


13/13 - 8s - 605ms/step - loss: 0.0629 - val_loss: 0.0622


Epoch 19/80


13/13 - 9s - 722ms/step - loss: 0.0612 - val_loss: 0.0587


Epoch 20/80


13/13 - 9s - 729ms/step - loss: 0.0587 - val_loss: 0.0567


Epoch 21/80


13/13 - 7s - 542ms/step - loss: 0.0567 - val_loss: 0.0549


Epoch 22/80


13/13 - 7s - 522ms/step - loss: 0.0550 - val_loss: 0.0536


Epoch 23/80


13/13 - 8s - 602ms/step - loss: 0.0537 - val_loss: 0.0531


Epoch 24/80


13/13 - 10s - 740ms/step - loss: 0.0525 - val_loss: 0.0511


Epoch 25/80


13/13 - 9s - 725ms/step - loss: 0.0512 - val_loss: 0.0500


Epoch 26/80


13/13 - 9s - 704ms/step - loss: 0.0507 - val_loss: 0.0501


Epoch 27/80


13/13 - 7s - 555ms/step - loss: 0.0497 - val_loss: 0.0483


Epoch 28/80


13/13 - 9s - 730ms/step - loss: 0.0484 - val_loss: 0.0478


Epoch 29/80


13/13 - 12s - 890ms/step - loss: 0.0476 - val_loss: 0.0471


Epoch 30/80


13/13 - 9s - 683ms/step - loss: 0.0469 - val_loss: 0.0464


Epoch 31/80


13/13 - 9s - 677ms/step - loss: 0.0465 - val_loss: 0.0463


Epoch 32/80


13/13 - 8s - 597ms/step - loss: 0.0456 - val_loss: 0.0448


Epoch 33/80


13/13 - 8s - 601ms/step - loss: 0.0448 - val_loss: 0.0443


Epoch 34/80


13/13 - 7s - 557ms/step - loss: 0.0445 - val_loss: 0.0442


Epoch 35/80


13/13 - 6s - 489ms/step - loss: 0.0445 - val_loss: 0.0438


Epoch 36/80


13/13 - 8s - 596ms/step - loss: 0.0434 - val_loss: 0.0430


Epoch 37/80


13/13 - 6s - 499ms/step - loss: 0.0426 - val_loss: 0.0427


Epoch 38/80


13/13 - 6s - 476ms/step - loss: 0.0420 - val_loss: 0.0419


Epoch 39/80


13/13 - 7s - 512ms/step - loss: 0.0415 - val_loss: 0.0416


Epoch 40/80


13/13 - 7s - 512ms/step - loss: 0.0413 - val_loss: 0.0417


Epoch 41/80


13/13 - 6s - 481ms/step - loss: 0.0409 - val_loss: 0.0407


Epoch 42/80


13/13 - 6s - 467ms/step - loss: 0.0404 - val_loss: 0.0408


Epoch 43/80


13/13 - 7s - 504ms/step - loss: 0.0400 - val_loss: 0.0403


Epoch 44/80


13/13 - 7s - 512ms/step - loss: 0.0403 - val_loss: 0.0402


Epoch 45/80


13/13 - 7s - 532ms/step - loss: 0.0397 - val_loss: 0.0400


In [6]:
import matplotlib.pyplot as plt

plt.figure(figsize=(6, 4))
plt.plot(history.history['loss'], label='train')
plt.plot(history.history['val_loss'], label='validation')
plt.xlabel('époque')
plt.ylabel('MSE')
plt.title("Courbe d'apprentissage")
plt.legend()
plt.tight_layout()
plt.show()

/var/folders/7v/dfg0p2vj1z58n65d99l2n1lh0000gp/T/ipykernel_18743/1096146220.py:11: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [7]:
from src.vision.visualization import plot_image_grid

n_examples = 5
reconstructions = autoencoder.predict(X_train_norm[:n_examples], verbose=0)

rows = [
    [X_train_norm[i, ..., 0] for i in range(n_examples)],
    [reconstructions[i, ..., 0] for i in range(n_examples)],
]
plot_image_grid(rows, 'Reconstructions sur des saines (train)', row_labels=['original', 'reconstruction'])

/Users/kenji/formation-ia/indusense_tp/src/vision/visualization.py:45: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


# Étape 3 — Score d'anomalie & seuil

In [8]:
from src.vision.autoencoder import reconstruction_score

VALIDATION_SPLIT = 0.2  # doit correspondre au validation_split utilisé dans le fit()

# Keras prend les derniers `validation_split` % de X_train_norm comme validation (avant shuffle) :
# on retrouve ce même sous-ensemble pour évaluer l'erreur de reconstruction.
split_at = int(len(X_train_norm) * (1 - VALIDATION_SPLIT))
X_val_norm = X_train_norm[split_at:]

reconstructions_val = autoencoder.predict(X_val_norm, verbose=0)
errors_val = reconstruction_score(X_val_norm, reconstructions_val)

print('Erreur de reconstruction (saines, validation) :', errors_val.shape)
print(f'  moyenne = {errors_val.mean():.5f}, écart-type = {errors_val.std():.5f}')

Erreur de reconstruction (saines, validation) : (52,)
  moyenne = 0.00726, écart-type = 0.00117


In [9]:
from src.vision.visualization import plot_score_histogram

threshold_p95 = np.percentile(errors_val, 95)

plot_score_histogram(
    errors_val, threshold_p95,
    title="Distribution de l'erreur de reconstruction (saines, validation)",
    xlabel='erreur de reconstruction (MSE)', threshold_label='seuil (centile 95)', precision=5,
)

/Users/kenji/formation-ia/indusense_tp/src/vision/visualization.py:16: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [10]:
reconstructions_test_good = autoencoder.predict(test_good_norm, verbose=0)
errors_test_good = reconstruction_score(test_good_norm, reconstructions_test_good)

print('Erreur de reconstruction (saines, test) :', errors_test_good.shape)
print(f'  moyenne = {errors_test_good.mean():.5f}, écart-type = {errors_test_good.std():.5f}')

Erreur de reconstruction (saines, test) : (41,)
  moyenne = 0.00567, écart-type = 0.00080


In [11]:
threshold_p95_test = np.percentile(errors_test_good, 95)

plot_score_histogram(
    errors_test_good, threshold_p95_test,
    title="Distribution de l'erreur de reconstruction (saines, test)",
    xlabel='erreur de reconstruction (MSE)', threshold_label='seuil (centile 95, test saines)', precision=5,
)

Courbes saines de validation et de tests sensiblement identiques 

In [12]:
reconstructions_test_defects = autoencoder.predict(test_defects_norm_stacked, verbose=0)
errors_test_defects = reconstruction_score(test_defects_norm_stacked, reconstructions_test_defects)

print('Erreur de reconstruction (défauts, test) :', errors_test_defects.shape)
print(f'  moyenne = {errors_test_defects.mean():.5f}, écart-type = {errors_test_defects.std():.5f}')

Erreur de reconstruction (défauts, test) : (119,)
  moyenne = 0.00778, écart-type = 0.00098


In [13]:
plot_score_histogram(
    errors_test_defects, threshold_p95_test, color='tab:orange',
    title="Distribution de l'erreur de reconstruction (défauts, test)",
    xlabel='erreur de reconstruction (MSE)', threshold_label='seuil (centile 95, test saines)', precision=5,
)

# Étape 4 — Heatmaps & évaluation

On prend trois images du set test_defects_norm_stacked et on affiche :
original | reconstruction | heatmap | masque ground-truth

In [14]:
from src.vision.data import build_defect_index, list_defect_types, load_mask

TEST_DIR = DATA_DIR / 'test'
GT_DIR = DATA_DIR / 'ground_truth'
defect_types = list_defect_types(TEST_DIR)

# Reconstitue, pour chaque indice de test_defects_norm_stacked, la catégorie et le chemin du masque
# correspondants (même ordre de concaténation qu'en partie 1 : catégories triées, fichiers triés dans
# chaque catégorie).
category_per_index, mask_paths = build_defect_index(TEST_DIR, GT_DIR, defect_types)

indices = [0, 50, 100]  # trois défauts de catégories différentes (déjà en mémoire, pas de recalcul)

rows = []
for idx in indices:
    original = test_defects_norm_stacked[idx, ..., 0]
    reconstruction = reconstructions_test_defects[idx, ..., 0]
    heatmap = (original - reconstruction) ** 2
    mask = load_mask(mask_paths[idx], IMG_SIZE)
    rows.append([original, reconstruction, heatmap, mask])

col_titles = ['original', 'reconstruction', 'heatmap', 'masque ground-truth']
plot_image_grid(
    rows, 'Défauts : original | reconstruction | heatmap | masque ground-truth',
    col_titles=col_titles, cmaps=['gray', 'gray', 'inferno', 'gray'],
)

# Évalution AUC-ROC

In [15]:
from src.vision.evaluation import plot_roc_and_confusion

y_true = np.concatenate([np.zeros(len(errors_test_good)), np.ones(len(errors_test_defects))])
y_score = np.concatenate([errors_test_good, errors_test_defects])

roc_auc, y_pred = plot_roc_and_confusion(
    y_true, y_score, threshold_p95_test, threshold_label='seuil centile 95, test saines', precision=5,
)

/Users/kenji/formation-ia/indusense_tp/src/vision/evaluation.py:22: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()
/Users/kenji/formation-ia/indusense_tp/src/vision/evaluation.py:30: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


### quel défaut est loupé ?

In [16]:
from src.vision.evaluation import missed_defect_counts, plot_missed_defects

n_good = len(errors_test_good)
counts, fn_idx = missed_defect_counts(y_true, y_pred, category_per_index, n_good)
plot_missed_defects(counts, defect_types)

manipulated_front   : 11 loupé(s)
scratch_head        : 3 loupé(s)
scratch_neck        : 6 loupé(s)
thread_side         : 7 loupé(s)
thread_top          : 3 loupé(s)


/Users/kenji/formation-ia/indusense_tp/src/vision/evaluation.py:53: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


15 prédites saines alors qu'avec défaut (mais 104 défaut trouvés correctement)
2 prédites en défauts alors que saines (mais 39 saines trouvés correctement)

# Test visuel pour comprendre

On prend les deux faux-positifs en défaut (alors que saine), et deux faux-positifs prédites saines sur les 15 (alors qu'en défaut)


In [17]:
n_good = len(errors_test_good)  # offset : y_true/y_pred = [saines de test, puis défauts de test]

fp_idx = np.where((y_true == 0) & (y_pred == 1))[0]          # saine → défaut, indices directs dans test_good_norm
fn_idx = np.where((y_true == 1) & (y_pred == 0))[0] - n_good  # défaut → saine, indices dans test_defects_norm_stacked

selected_fp = fp_idx[:2]
selected_fn = fn_idx[:2]

# (étiquette, indice, images, reconstructions, chemins des masques ou None si pas de défaut à localiser)
cases = [('faux positif : saine → défaut', i, test_good_norm, reconstructions_test_good, None) for i in selected_fp]
cases += [('faux négatif : défaut → saine', i, test_defects_norm_stacked, reconstructions_test_defects, mask_paths) for i in selected_fn]

rows, row_labels = [], []
for label, idx, images, recons, masks in cases:
    original = images[idx, ..., 0]
    reconstruction = recons[idx, ..., 0]
    heatmap = (original - reconstruction) ** 2
    mask = load_mask(masks[idx], IMG_SIZE) if masks is not None else np.zeros_like(original)
    rows.append([original, reconstruction, heatmap, mask])
    row_labels.append(label)

col_titles = ['original', 'reconstruction', 'heatmap', 'masque ground-truth']
plot_image_grid(
    rows, 'Erreurs de classification (faux positifs et faux négatifs)',
    col_titles=col_titles, row_labels=row_labels, cmaps=['gray', 'gray', 'inferno', 'gray'],
)

/Users/kenji/formation-ia/indusense_tp/src/vision/visualization.py:45: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


et la zone de 16x16 à 8x8, quel impact pour le manipulated_front ?

Ça dépend de la nature du défaut manipulated_front : s'il s'agit d'une altération globale de la tête (orientation, forme de l'empreinte) plutôt qu'une texture locale, passer de 16×16 à 8×8 force chaque cellule du goulot à résumer une zone 2× plus grande (16×16 px au lieu de 8×8 px) — le modèle encoderait alors plus de forme/structure globale et moins de détail fin, ce qui pourrait mieux capter ce type d'altération. Le tradeoff : cette perte de résolution spatiale devrait pénaliser les défauts localisés et fins (scratch_head, scratch_neck, filetage), qui ont justement besoin de précision spatiale pour ne pas être noyés dans une moyenne plus grossière.

C'est une hypothèse plausible mais pas garantie sans test — je n'ai pas de certitude sur ce que manipulated_front représente visuellement ici. 